include/eosio/vm/host_function.hpp

```diff
@@ -214,24 +214,32 @@ namespace eosio { namespace vm {
    template <typename T, typename WAlloc>
    constexpr auto resolve_result(T&& res, WAlloc*) -> std::enable_if_t<
          !(std::is_pointer_v<T> || std::is_reference_v<T>)&&std::is_same_v<to_wasm_t<T>, i32_const_t>, i32_const_t> {
-      return i32_const_t{ *(uint32_t*)&res };
+      if constexpr (std::is_same_v<std::decay_t<T>, bool>) {
+         return i32_const_t{ res };
+      } else {
+         static_assert(sizeof(T) == sizeof(uint32_t));
+         return i32_const_t{ *(uint32_t*)&res };
+      }
```

Comment:
This shouldn't be using pointer casts in the first place.  It should just use `static_cast<uint32_t>(res)` for all types.